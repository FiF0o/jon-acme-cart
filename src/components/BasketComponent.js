import React from 'react';
import PropTypes from 'prop-types';

import { ProductList } from './ProductList'


const BasketComponent = ({ basket, getData, isLoading, hasError }) =>
  <div className="mdc-layout-grid mdc-layout-grid">
    <div className="mdc-layout-grid__cell--span-12">
      <h1>{`Your basket contains ${basket.length} item(s)`}</h1>

    </div>
    <div className="mdc-layout-grid__inner">
      <div className={`mdc-layout-grid__cell--span-${!basket.length > 1 ? '12' : '6'}`}>
        { hasError &&
          <h3 className="mdc-typography--headline3 has-error">It seems that something went wrong <span role="img" aria-label="explosion">ðŸ’¥</span></h3> }
        {
          isLoading &&
            <div role="progressbar" className="mdc-linear-progress">
              <div className="mdc-linear-progress__buffering-dots"></div>
            </div>
        }
        {
          basket.length > 1 ?
            <ProductList basket={basket} /> :
            <div className="expand-full">
              <h2 className="sub-heading-title">
              Buy mooar it is empty <span role="img" aria-label="ouch emoji">ðŸ¤¯</span>
              </h2>
              &nbsp;
              <button
                className="mdc-button mdc-button--outlined"
                onClick={getData}
              >
                Load cart
              </button>
            </div>
        }
      </div>
      {
        basket.length > 1 &&
        <div className="mdc-layout-grid__cell--span-6">
          <button
            className="mdc-button mdc-button--raised"
          >
            Go to checkout
          </button>
          &nbsp;
          <button
            disabled={true}
            className="mdc-button mdc-button--outlined"
          >
            Emty basket
          </button>
        </div>
      }
    </div>
  </div>

BasketComponent.propTypes = {
  basket: PropTypes.array.isRequired,
  getData: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
  hasError: PropTypes.bool.isRequired
}

export {
  BasketComponent
}