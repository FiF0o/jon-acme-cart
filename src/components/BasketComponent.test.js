import React from 'react'
import { mount, shallow } from 'enzyme'

import { BasketComponent } from './BasketComponent'



describe('<BasketComponent /> component', () => {
  let wrapper, props;

  describe('without items', () => {
    beforeEach(() => {
      props = {
        basket: [],
        getData: jest.fn(),
        isLoading: false,
        hasError: false,
      }

      wrapper = mount(
        <BasketComponent {...props} />
      )
    })

    afterEach(() => {
      jest.clearAllMocks()
    })

    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should render a spinner when isLoading is true', () => {
      expect(wrapper.find('.mdc-linear-progress').length).toEqual(0)
      wrapper.setProps({ isLoading: true })
      expect(wrapper.find('.mdc-linear-progress').length).toEqual(1)
    })

    it('should render an error when hasError is true', () => {
      expect(wrapper.find('.has-error').length).toEqual(0)
      wrapper.setProps({ hasError: true })
      expect(wrapper.find('.has-error').length).toEqual(1)
    })


    it('should call getData() when the button is clicked', () => {
      const mock = props.getData
      expect(mock).not.toHaveBeenCalled()
      // getData() is the only button present on the page when the cart is empty
      wrapper.find('button').simulate('click')
      expect(mock).toHaveBeenCalled()
    })
  })

  describe('with items', () => {
    it('should render a list of items when baskt contains items', () => {
      const props = {
        basket: [{}, {}],
        getData: () => {},
        isLoading: false,
        hasError: false
      }
      const wrapper = shallow(<BasketComponent {...props} />)
      expect(wrapper.find('ProductList').length).toBe(1)
    })
  })

})