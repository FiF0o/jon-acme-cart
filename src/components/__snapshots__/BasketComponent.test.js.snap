// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BasketComponent /> component without items should render correctly 1`] = `
<div
  className="mdc-layout-grid mdc-layout-grid"
>
  <div
    className="mdc-layout-grid__cell--span-12"
  >
    <h1>
      Your basket contains 0 item(s)
    </h1>
  </div>
  <div
    className="mdc-layout-grid__inner"
  >
    <div
      className="mdc-layout-grid__cell--span-6"
    >
      <div
        className="expand-full"
      >
        <h2
          className="sub-heading-title"
        >
          Buy mooar it is empty 
          <span
            aria-label="ouch emoji"
            role="img"
          >
            ðŸ¤¯
          </span>
        </h2>
        Â 
        <button
          className="mdc-button mdc-button--outlined"
          onClick={[MockFunction]}
        >
          Load cart
        </button>
      </div>
    </div>
  </div>
</div>
`;
